### DISCUSSION TITLE
Is there anyway to get hostname or computername of the virtual machines?

### DISCUSSION THREAD

**Discussion by ngocnd1** [netbox-community/netbox#12220] [Q&A] [answered]
Hi Everyone,
Is there anyway to get hostname or computername of the virtual machines?
Name of Virtual Machines and hostname or computername are differences.
And if has how i can find it in dashboard?
Many Thanks!

**Comment by candlerb** [comment]
When you create a virtual machine in Netbox, you give it a name.  That's whatever name you choose that makes sense to you, using whatever naming convention you want.  As far as Netbox is concerned, that *is* the "name" of the virtual machine, by definition.

What other kind of "name of virtual machine" or "hostname" or "computer name" are you referring to?

**Reply by candlerb** [reply, marked as answer]
You are using Netbox in a non-recommended way: Netbox is not meant to be a portal onto data in other systems. Please see https://docs.netbox.dev/en/stable/introduction/#serve-as-a-source-of-truth where it says:

"automated import of live network state is strongly discouraged"

In other words: if you want to see what VMs you have in vCenter and what state they are in, then you should go to the vCentre management interface. That's what it's there for.

Of course, Netbox is free and open source and you are free to use it in any way that you like; but in that case, I suggest you raise your problem with whatever third-party sync tool you are using, not with Netbox itself.

**Reply by ngocnd1** [reply]
Thansk Candlerb,

I will try to using it in recommended ways, may be import via some ways to custom field.

**Reply by zmaster7** [reply]
If you are looking to pull the guest OS name from vCenter, you are looking for the Config.HostName property using Get-VM or Get-View PowerCLI Cmdlets. In theory, that DNS/hostname would be bound to one or more IP addresses (since that name is "worthless" otherwise). Therefore, in NetBox that name would be associated with the IP address(es) assigned to that VM, under the "DNS Name" field. At least that is how I went about it in my environment.

As candlerb mentioned however, it's not recommended to completely automate the import of data unless you can verify that the data matches what the state "should be" not "what it is currently." What I do in my environment is export the vCenter information to a JSON file using a custom PowerCLI script to verify everything looks good, then I use another PowerShell script to use the NetBox API to create/update/delete information as needed (with confirmations along the way).

Hope this helps.

**Reply by candlerb** [reply]
FYI, the upcoming Netbox v3.5 will have a [framework](https://github.com/netbox-community/netbox/blob/feature/docs/release-notes/version-3.5.md#remote-data-sources-11558) which potentially will make accessing data from remote data sources more straightforward, e.g. for the reconciliation use case.

**Reply by ngocnd1** [reply]
@zmaster7  Thanks for your solution, i will try soon in my dev environment.

**Reply by ngocnd1** [reply]
> FYI, the upcoming Netbox v3.5 will have a [framework](https://github.com/netbox-community/netbox/blob/feature/docs/release-notes/version-3.5.md#remote-data-sources-11558) which potentially will make accessing data from remote data sources more straightforward, e.g. for the reconciliation use case.

Wow, interesting.

**Reply by zmaster7** [reply]
> @zmaster7 Thanks for your solution, i will try soon in my dev environment.

I should add, if you are looking to match on existing VMs within NetBox, I did create a custom field for vm_uuid and I pull Config.Uuid from vCenter to populate that data. When I'm looking for a match, I first check for that Uuid, then by VM Name (vCenter-assigned name) and Cluster Name. If neither of those routes find a match, then it is usually safe (after verifying) to create a new VM in NetBox instead.
